File: Chapter.14/StockTicker.Server.CSharp/Bus/CommandHandler.cs
using System;
using System.Threading.Tasks;
using StockTicker.Core;
using StockTicker.Server.CSharp.Core;
using static StockTicker.Core.Models;

namespace StockTicker.Server.CSharp.Bus
{
    public static class CommandHandler
    {
        private static readonly TradingCoordinator TradingCoordinator = TradingCoordinator.Instance();
        private static readonly EventStorage.EventStorage Storage = EventStorage.EventStorage.Instance();

        public static async Task TradingCoordinatorHandle(CommandWrapper commandWrapper)
        {
            var connectionId = commandWrapper.ConnectionId;

            var command = TradingSupervisorAgent.CoordinatorMessage.NewPublishCommand(connectionId, commandWrapper);
            await TradingCoordinator.PublishCommand(command); // #E
        }

        public static void EventStorageHandle(CommandWrapper commandWrapper)
        {   
            var eventData = default(Events.Event);
            switch (commandWrapper.Command)
            {
                case TradingCommand.BuyStockCommand cmdBuy:
                    eventData = Events.Event.NewStocksBoughtEvent(commandWrapper.Id, cmdBuy.tradingRecord);
                    break;
                case TradingCommand.SellStockCommand cmdSell:
                    eventData = Events.Event.NewStocksSoldEvent(commandWrapper.Id, cmdSell.tradingRecord);
                    break;
            }

            var eventDescriptor = Events.Event.Create(commandWrapper.Id, eventData);
            Storage.SaveEvent(commandWrapper.Id, eventDescriptor);
        }
    }
}

File: Chapter.14/StockTicker.Server.FSharp/Bus/CommandHandler.fs
namespace StockTicker.Server.FSharp.Bus

open System
open PublishWithRetry
open StockTicker.Server.FSharp
open TradingCoordinator
open StockTicker.Core
open Events
open EventStorage

module private Retry =
    // Listing 9.4 Retry Async Builder
    let retryPublish = RetryAsyncBuilder(10, 250) // #A

// Listing 14.4 Command handler with async retry logic
module CommandHandler =
    open Retry
    
    let tradingCoordinator = TradingCoordinator.Instance()   // #B
    
    let AsyncHandle (commandWrapper:CommandWrapper) =   // #D
        let connectionId = commandWrapper.ConnectionId

        retryPublish {      // #A
            tradingCoordinator.PublishCommand(CoordinatorMessage.PublishCommand(connectionId, commandWrapper))   // #E
    }
        
// Listing 14.4 Command handler with async retry logic
module EventStorageHandler =
    open Retry
    let Storage = new EventStorage()     // #C

    let AsyncHandle (commandWrapper:CommandWrapper) =   // #D
        let connectionId = commandWrapper.ConnectionId

        retryPublish {      // #A
            let event =
                let cmd = commandWrapper.Command
                match cmd with     // #F
                | BuyStockCommand(connId,trading) -> StocksBoughtEvent(commandWrapper.Id, trading)
                | SellStockCommand(connId, trading) -> StocksSoldEvent(commandWrapper.Id, trading)   // #F

            let eventDescriptor = Event.Create (commandWrapper.Id, event)
            Storage.SaveEvent (Guid(connectionId), eventDescriptor)    // #G
        }        